Avoid exceptions when console is not available

pharring opened the pull request:
Fixes #637

WojciechNagorski commented:
I think it is not a good solution. Could you try to do it in this way:
```
if (GetConsoleWindow() != IntPtr.Zero) Console.Clear();
```

Links:
https://social.msdn.microsoft.com/Forums/vstudio/en-US/074a032d-5268-4c9b-9de9-2c2464215677/calling-consoleclear-in-a-console-app-within-another-net-app-through-process-class?forum=netfxbcl

https://www.pinvoke.net/default.aspx/kernel32.getconsolewindow

There is my issue #645

pharring commented:
It needs to work on non-Windows platforms too, so I can't P/Invoke to Windows-only APIs.

WojciechNagorski commented:
Sorry you're right. https://docs.microsoft.com/en-us/dotnet/api/system.console.clear?view=netframework-4.8#remarks

jorive commented on src/Tools/dotnet-trace/CommandLine/Commands/CollectCommand.cs:
```diff
@@ -38,7 +38,16 @@ private static async Task<int> Collect(CancellationToken ct, IConsole console, i
+                }
+                catch (IOException)
+                {
+                    // Console not available.
```
nit: Should we add a link to documentation here, or explanation?

pharring replied:
Fixed in 2nd iteration

pharring commented:
@jorive I added a link to the doc for System.Console.Clear().
@josalem You are correct. I updated this PR to account for SetCursorPosition too. The output is not pretty (when the trace ends you get a slew of output as buffers are flushed), but at least it succeeds. I tested this version by launching it from a wrapper app that redirects stdout/stderr. Testing on Azure App Service for #637 is currently blocked by an UnauthorizedAccessException on the diagnostics named pipe. I'm working with the App Service team on a fix for that.

josalem commented on src/Tools/dotnet-trace/CommandLine/Commands/CollectCommand.cs:
```diff
@@ -38,7 +38,20 @@ private static async Task<int> Collect(CancellationToken ct, IConsole console, i
+                {
+                    // The console is not available, probably because it has been redirected.
+                    // See https://docs.microsoft.com/dotnet/api/system.console.clear#remarks
+                    hasConsole = false;
```
While this works, I would prefer that we use the S.CommandLine way of checking for a console and/or preemptively checking if we have a console rather than catching the exception, e.g., 

This will only clear the terminal if we have a terminal
```csharp
console.GetTerminal()?.Clear(); // note this uses the IConsole console and not System.Console
```

This will check if we have a terminal to clear/rewrite the line in:
```csharp
if (console.GetTerminal() is ITerminal terminal)
{
    // use terminal here to do things like set cursor position
    // can also use System.Console here since we know we have one
}
```
or simply
```csharp
bool hasConsole = console.GetTerminal() != null;
```

pharring replied:
OK. I went with the last suggestion.

josalem commented:
@pharring I was taking another look at redirection and I noticed that this PR didn't wrap our command loop that uses `Console.KeyAvailable` and `Console.ReadKey(bool)` which _should_ be throwing under redirection.  Did you get `dotnet-trace` working in the Kudu console or are you still facing the exception?  When I run it locally with a redirected input, I still get the exception.

pharring commented:
@josalem You're probably right. I tested stdout and stderr redirection only using the code from https://github.com/dotnet/diagnostics/issues/645#issue-528128386 
As I mentioned, it doesn't work in Kudu because the app's named pipe isn't accessible from the SCM (where Kudu runs).
